### PR TITLE
feat(traits): Add default and override of methods

### PR DESCRIPTION
**Pull request by yordanmadzhunkov**
This is the first part from split  https://github.com/noir-lang/noir/pull/2574

#Description #2568
In this PR we introduce the following progress of implementation of traits:
- [x] Trait can have an default implementation that will execute if not override by the struct
- [x] Trait can have an empty implementation which forces the struct to provide one
- [x] Structs can implement traits and override the default implementation

Remove unnecessary checks:
 - [x] Remove the premature checks (from def collector), that should be done at a later phase:

Name resolution/Type checking:
 - [x] Check for duplicate implementations for the same Trait and Type (this check should work even for empty traits!) 
 - [x] Error when implementing non-struct type on trait

Successful examples:

`trait_default_implementation`
```
trait Default {
    fn default(x: Field, y: Field) -> Self;
    fn method2(x: Field) -> Field {
	x
    }
}

struct Foo {
    bar: Field,
    array: [Field; 2],
}

impl Default for Foo {
    fn default(x: Field,y: Field) -> Self {
        Self { bar: x, array: [x,y] }
    }
}

fn main(x: Field) {
    let first = Foo::method2(x);
    assert(first == x);
}
```

`trait_override_implementation`
```
trait Default {
    fn default(x: Field, y: Field) -> Self;
    fn method2(x: Field) -> Field {
	x
    }
}

struct Foo {
    bar: Field,
    array: [Field; 2],
}

impl Default for Foo {
    fn default(x: Field,y: Field) -> Self {
        Self { bar: x, array: [x,y] }
    }
    fn method2(x: Field) -> Field {	
        x * 3    
    }
}

fn main(x: Field) {
    let first = Foo::method2(x);
    assert(first == 3 * x);
}
```

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by yordanmadzhunkov** [comment]
@jfecher :)

File: crates/nargo_cli/tests/compile_failure/dup_trait_implementation_3/src/main.nr [hunk 1]
```diff
@@ -0,0 +1,19 @@
+trait Default {
+}
+
+struct MyStruct {
+}
+
```
Comment (jfecher) [inline crates/nargo_cli/tests/compile_failure/dup_trait_implementation_3/src/main.nr:6]: Do we really need `dup_trait_implementation_1`, 2, and 3 just to test duplicate traits? These look like they should all be covered by the same check to me, so there should only be one test.

File: crates/noirc_frontend/src/hir/def_collector/dc_crate.rs [hunk 2]
```diff
@@ -601,6 +660,55 @@ fn resolve_impls(
...
+            let mut resolver =
+                Resolver::new(interner, &path_resolver, &context.def_maps, trait_impl.file_id);
+            resolver.resolve_type(unresolved_type.clone())
+        };
+
+        let mut impl_methods = resolve_function_set(
+            interner,
+            crate_id,
+            &context.def_maps,
+            trait_impl.methods.clone(),
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir/def_collector/dc_crate.rs:687]: Can we remove the `#[derive(Clone)]` for unresolved functions and unresolved traits if we remove the clone here? I think `resolve_function_set` does not actually need to take its arguments by value.

File: crates/noirc_frontend/src/hir/def_collector/errors.rs [hunk 3]
```diff
@@ -29,6 +30,8 @@ pub enum DefCollectorErrorKind {
     PathResolutionError(PathResolutionError),
     #[error("Non-struct type used in impl")]
     NonStructTypeInImpl { span: Span },
+    #[error("Non-struct type used in trait impl")]
+    NonStructTraitImpl { trait_ident: Ident, span: Span },
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir/def_collector/errors.rs:34]: Non-struct types are allowed in trait impls. The trait should just be defined in the current crate if so.

File: crates/noirc_frontend/src/node_interner.rs [hunk 4]
```diff
@@ -71,6 +73,11 @@ pub struct NodeInterner {
     // We'd just lookup their methods as needed through the NodeInterner.
     traits: HashMap<TraitId, Shared<Trait>>,
 
+    // Trait implementation map
+    // For each type that implements a given Trait ( corresponding TraitId), there should be an entry here
+    // The purpose for this hashmap is to detect duplication of trait implementations ( if any )
+    trait_implementaions: HashMap<(Type, TraitId), Ident>,
```
Comment (jfecher) [inline crates/noirc_frontend/src/node_interner.rs:79]: I think we'll need to refine this further in the future to catch duplicate generic implementations. E.g. `impl<U> Foo for Bar<U>` and `impl<U> Foo for Bar<U>` are equivalent but would be different keys in this map.

**Review by jfecher — APPROVED** [review]
There's a few items here that can be cleaned up (see comments), but we can leave those to a future PR. I'll be on vacation until Monday the 18th starting tomorrow so I'd like to approve this now to get it merged.

While I'm gone, feel free to keep making PRs. You can set the merge target of each successive PR to the branch in the previous PR so that you can keep building on the code even if the prior PR(s) aren't yet merged. Might have some merge conflicts when changes are made to an earlier PR later but I think this would be better than halting all progress for a week. Apologies, I know this is suboptimal!

File: crates/noirc_frontend/src/node_interner.rs (hunk 4 above)
Comment (alexvitkov) [reply crates/noirc_frontend/src/node_interner.rs:79]: Yeah, in my local branch I'm currently implementing generic traits, the signature looks like:
```rust
struct TraitImpl {
    pub methods: Vec<FuncId>, // list of the concrete method implementation
    ...
}

struct TraitImplKey {
    typ: Type,
    trait_id: TraitId,
    generics: Vec<Type<
}

trait_implementaions: HashMap<(Type, TraitId), Shared<TraitImpl>>,

File: crates/noirc_frontend/src/hir/def_collector/errors.rs (hunk 3 above)
Comment (alexvitkov) [reply crates/noirc_frontend/src/hir/def_collector/errors.rs:34]: We've added this to the roadmap, https://github.com/noir-lang/noir/issues/2568

The issue preventing us from allowing this right away, is that structs have their own module, in which the trait methods get declared, which I believe is not the case for say primitive types, or typedefs

File: crates/nargo_cli/tests/compile_failure/dup_trait_implementation_3/src/main.nr (hunk 1 above)
Comment (alexvitkov) [reply crates/nargo_cli/tests/compile_failure/dup_trait_implementation_3/src/main.nr:6]: They're named a bit poorly, but `dup_trait_implementation_1` tests duplicate methods inside the same impl block, and `2` tests for duplicate impl blocks. `3` is similar to `2`, but it explicitly uses a typedef for the struct, I'd rather keep it in since we had a few troubles around them.

File: crates/noirc_frontend/src/hir/def_collector/dc_crate.rs (hunk 2 above)
Comment (alexvitkov) [reply crates/noirc_frontend/src/hir/def_collector/dc_crate.rs:687]: There's a few structures which we clone very liberally, which I believe we can avoid cloning at all - mainly NoirFunction (which is 300 bytes + has a bunch of Vecs) as well as the AST nodes for Trait/TraitImpl.

Doing this would require changing quite a few method signatures, so it's probably a PR of its own, and we've already got a few PRs planned, so if you want I can try to do it after we get the following merged:

1. Trait impl methods signature checks PR: https://github.com/noir-lang/noir/pull/2652
2. Trait Bounds DC+Resolver PR: 
3. Trait Bounds Typechecker+Monomorphizer PR:
